src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -615,27 +588,32 @@ private async Task ComputeAnalyzerDiagnosticsAsync(AnalysisScope? analysisScope,
 
             Debug.Assert(analysisScope != null);
 
-            var originalAnalyzers = analysisScope.Analyzers;
-            Compilation compilation;
-            AnalyzerDriver? driver = null;
-            bool driverAllocatedFromPool = false;
-
             try
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
-                analysisScope = GetPendingAnalysisScope(analysisScope, _analysisResultBuilder);
+                analysisScope = GetPendingAnalysisScope(analysisScope);
                 if (analysisScope == null)
                     return;
 
-                (compilation, driver, driverAllocatedFromPool) = await getCompilationAndDriverAsync().ConfigureAwait(false);
+                // Get the compilation to execute analysis.
+                // PERF: We reuse the root "_compilation" when computing single-file diagnostics only for the compiler analyzer.
+                // For rest of the cases, we use a cloned compilation with a new event queue and semantic model provider
```

Comment:
I'm confused by this. If we need a cloned compilation with more specific semantic model, I would expect it to be for the single-file diagnostics case rather than the (full compilation?) case. Do we end up here multiple times for distinct single-file cases, and if so, do those cases all use the same compilation?

Reply:
>  I would expect it to be for the single-file diagnostics case rather than the other (full compilation?) case. Do we end up here multiple times for distinct single-file cases

@cston We used to share the same compilation for different single file diagnostic requests earlier, but this required us to do partial analysis state tracking for the compilation- we need to track which analyzer has received callback for which node/operation/symbol across calls into CompilationWithAnalyzers to ensure we don’t make duplicate callbacks into an analyzer for the same compilation. However, we identified with performance measurements that this was causing very high memory usage as well as higher execution times from locking/synchronization code for this state management. After removing the partial state tracking, we use separate  cloned compilation for each single file request, avoiding us to have to do this state tracking and allowing us to discard the cloned compilation after analyzing that file. However, we can still re-use the same compilation for single file requests with just the compiler analyzer being a single analyzer in the request (IDE computes and refreshes compiler syntax and semantic diagnostics upfront before dealing with analyzer diagnostics). We know this sharing model works fine as CompilerAnalyzer is just making calls into SyntaxTree.GetDiagnostics and SemanticModel.GetDiagnostics on analyzer callbacks and duplicate callbacks are not problematic.